packages/mobile/src/account/profileInfo.ts

```diff
@@ -0,0 +1,317 @@
+// and should be avoided for large files!
+function blobToBuffer(blob: Blob) {
+  const reader = new FileReader()
+  reader.readAsDataURL(blob)
+  return new Promise<Buffer>((resolve, reject) => {
+    reader.onerror = () => {
+      reject(reader.error)
+    }
+    reader.onloadend = () => {
+      const result = reader.result
+      if (typeof result !== 'string') {
+        reject(new Error('Unexpected result type'))
+      } else {
+        // Result looks like "data:application/octet-stream;base64,BLMHTkM..."
+        // Extract the base64 part
+        const base64 = result.substr(result.lastIndexOf(',') + 1)
+        resolve(Buffer.from(base64, 'base64'))
+      }
+    }
+  })
+}
+
+class UploadServiceDataWrapper implements OffchainDataWrapper {
+  signer: Address
+  self: Address
+
+  constructor(readonly kit: ContractKit, address: Address) {
+    this.signer = this.self = address
+  }
+
+  async writeDataTo(
+    data: Buffer,
+    signature: Buffer,
+    dataPath: string
+  ): Promise<OffchainErrors | void> {
+    const dataPayloads = [data, signature]
+    const signedUrlsPayload = [
+      {
+        path: dataPath,
+      },
+      {
+        path: `${dataPath}.signature`,
+      },
+    ]
+
+    const hexPayload = ensureLeading0x(
+      Buffer.from(JSON.stringify(signedUrlsPayload)).toString('hex')
+    )
+    const authorization = await this.kit.getWallet().signPersonalMessage(this.signer, hexPayload)
+    const signedUrls = await makeCall(signedUrlsPayload, authorization)
+    await Promise.all(
+      signedUrls.map(({ url, fields }, i) => {
+        const formData = new FormData()
+        for (const name of Object.keys(fields)) {
+          formData.append(name, fields[name])
+        }
+        formData.append('file', dataPayloads[i])
+
+        return fetch(url, {
+          method: 'POST',
+          headers: formData.getHeaders(),
+          // Use getBuffer() which ends up transferring the body as base64 data over the RN bridge
+          // because RN doesn't support Buffer inside FormData
+          // See https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/convertRequestBody.js#L34
+          // and https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/FormData.js
+          body: formData.getBuffer(),
+        }).then((x) => {
+          if (!x.ok) {
+            Logger.error(TAG + '@writeDataTo', 'Error uploading ' + x.headers.get('location'))
+          }
+          return x.text()
+        })
+      })
+    )
+  }
+
+  async readDataFromAsResult<DataType>(
+    account: Address,
+    dataPath: string,
+    _checkOffchainSigners: boolean,
+    type?: t.Type<DataType>
+  ): Promise<Result<Buffer, OffchainErrors>> {
+    let dataResponse, signatureResponse
+
+    const accountRoot = `${config.CIP8MetadataUrl}/${toChecksumAddress(account)}`
```

Comment:
This is an optimisation for Valora users, we should fallback to the usual offchain read function if it fails. 

We should also consider registering this metadata URL for Valora users, otherwise only other Valora users will know how to resolve ther name and profile picture. If they have already set their metadataURL we can consider aborting or overwriting it, I'm not sure which is the right solution.

Reply:
do you think this should be implemented in this PR? I was thinking it would be added in later

Reply:
I think we can get away without it in this PR yeah. Should be top priority though once this is shipped and stable.

cc @aslawson

Reply:
Realized I didn't publish this yesterday, so just published to document thoughts.

Talked with @AlexBHarley and @nambrot and it sounds like current state is that this is calling the storage url directly, so there is additional work still to create the metadata blobs.

I think it's ok as fast follow if Valora provides a way for users to later "share to wider network", so there is a route for them those to register on-chain if they add their photos/names before we have that implemented.